CSV Sniffer - State Machine

pdet opened the pull request:
I'm genuinely sorry for the massive PR. I swear I've tried to make it as minimal as possible.

# Featured Changes:
## CSV State Machine.
We now use a CSV State machine (as described in #7213). The state machine is generated based on the options set (e.g., delimiter, quote, ...) for that CSV Read. It is currently only used in the Sniffer and implements operations that parse a CSV File.

**FW:** There is still significant branching that can be removed from the existing parsing functions of the state machine. We can return more efficient types (than values and data chunks) and have specialized try_cast methods.

## CSV Buffer Manager.
During the sniffing and on the initial runs of the actual CSV Parsing, one or more CSV Buffers will be cached and properly reused. I've created a buffer manager class that manages, caches and removes buffers accordingly. One thing to notice is that the char `CSVBufferIterator::GetNextChar()` function is very inefficient.
**FW:** This should be rewritten to avoid all the branching of buffer checking.

## New Sniffer Code.
The previous sniffer code was quite ingrained with the Buffered CSV Reader (Our single-threaded CSV Reader implementation). I've separated it from it and created a `CSVSniffer` class that performs the sniffing. Besides that, the Sniffer now runs on `options.sample_chunks`. Before dialect detection would only run on one chunk, this should fix errors like #7789.
The main goal of sniffing is to detect column types, column names, and the CSV options used to parse a CSV File.
In summary, it consists of the following steps:

### Dialect Detection
Generate the CSV Options (delimiter, quote, escape, etc.)
Dialect Detection consists of four phases:
1)  Generates the search space candidates for the dialect. In this phase, we generate the search space of options for this CSV File. This is based on our default (or user-defined) options for quotes, escapes, and delimiters.
2) Generate State Machines: After defining our search space, we generate one CSV State Machine for each combination of quotes, escape, and delimiter.
3) Analyze Dialect Candidates: For each state machine created in phase 2, we will parse the first chunk of the CSV File and determine, based on the number of consistent rows and the number of columns, our top State Machine candidates.
4) Refine Candidates: It repeats step 3 over the remaining chunks set on `options.sample_chunks`. Eliminating candidates with inconsistent rows.

After Running the dialect detection method, we have our best candidates (i.e., the ones with the most consistent rows and the maximum number of columns) from our search space.

### Type Detection
Figures out the types of columns (For the first chunk).

For each row of the first chunk, it tries to cast columns defined in our `auto_type_candidates` variable.
Since we don't know if this file has a header, we eliminate the first row and consider it a possible header. This will be verified in the next phase.
```cpp
//! Types considered as candidates for auto-detection ordered by descending specificity (~ from high to low)
	vector<LogicalType> auto_type_candidates = {LogicalType::VARCHAR, LogicalType::TIMESTAMP, LogicalType::DATE,
	                                            LogicalType::TIME,    LogicalType::DOUBLE,    LogicalType::BIGINT,
	                                            LogicalType::BOOLEAN, LogicalType::SQLNULL};
```
It starts with the `back()` type and pops it in case it can't cast it until finding the first elements type that works. It then continues the same process for the remaining rows.
We also try to detect their format for `Date` and `Timestamp` types, following a similar logic.
```cpp
	//! Format Candidates for Date and Timestamp Types
	const std::map<LogicalTypeId, vector<const char *>> format_template_candidates = {
	    {LogicalTypeId::DATE, {"%m-%d-%Y", "%m-%d-%y", "%d-%m-%Y", "%d-%m-%y", "%Y-%m-%d", "%y-%m-%d"}},
	    {LogicalTypeId::TIMESTAMP,
	     {"%Y-%m-%d %H:%M:%S.%f", "%m-%d-%Y %I:%M:%S %p", "%m-%d-%y %I:%M:%S %p", "%d-%m-%Y %H:%M:%S",
	      "%d-%m-%y %H:%M:%S", "%Y-%m-%d %H:%M:%S", "%y-%m-%d %H:%M:%S"}},
	};
```

We run the type detection for each candidate from the previous phase. We keep the one that returns the least `LogicalType::VARCHAR` columns as our best candidate.

**FW:** Value Casting could be more efficient. I imagine we can have specialized code to try to cast `char*` directly.

### Header Detection
 Figures out if the CSV file has a header and produces the names of the columns.
We try to cast all columns from our possible header to the type detection we did in the previous phase. If they don't match, we consider the first row a header.

### Type Replacement: 
Replaces the types of columns if the user specified them. Since these are supplied with a column_name:column_type map, we must first run the header detection to know the column names.

### Type Refinement: 
Refines the types of columns for the remaining chunks.
We continue the type-detection process for the remaining `options.sample_chunks`. The main difference between this phase to the Type Detection phase is that we use vector casts instead of value casts. We also don't try to continue `date` and `time` format detection.
**FW:** The same as Type detection should apply.

## Removal of complex CSV Options.
No more multi-character quotes, delimiters, and escapes. Most libraries do not support this. I've also removed tests related to that.


## Micro Regression Tests
I've added tests for the CSV Reader:
1. Sniffer.
Old timing: 0.107391
New timing: 0.164979

2. Small CSV Reader.
Old timing: 0.000259
New timing: 0.000612

3. CSV Reader
Old timing: 7.447226
New timing: 6.886744

# Future Work:
1. Performance: As described above, there are many steps that can be optimized to ensure faster CSV Parsing.
2. Removal of Buffered CSV Reader: When executed with a single thread, the Parallel CSV should behave like the Buffered CSV Reader. 
3. Implement Sampling on Sniffer: We are currently doing sequential access on the CSV for sniffing; however, CSV files may change their column types or dialect over longer periods of time, hence being able to run the Sniffer over different parts of the file.
6. Use CSV State Machine in the Parallel CSV Reading parsing.

pdet committed:
Sketch of new sniffer

pdet committed:
Move CSV Reader code to its own folder

pdet committed:
Basics of state machine creation

pdet committed:
Starting integration of Dialect Sniffer with State Machine

pdet committed:
Move sniffer code out of buffered reader

pdet committed:
changing to branch

pdet committed:
Moving Sniffing methods to sniffing class

pdet committed:
Got CSV Reader to run

pdet committed:
Removing complex values

pdet committed:
Remove more of complex csv options

pdet committed:
More of removal of complex options

pdet committed:
More of the removing complex options

pdet committed:
Fixing predication for newline detection

pdet committed:
Properly handle quotes and escapes in the state machine

pdet committed:
Remove max_row from dialect detection, count columns from dialect detection from 1

pdet committed:
Removing tests and files related to multi byte csv options

pdet committed:
Skip rows on dialect detection if options are set

pdet committed:
Skip rows on dialect detection if options are set

pdet committed:
Properly add escapes when getting new values

pdet committed:
Auto detection now fails for the null padding tests since we check more than one chunk. Which makes it discard all dialect candidates when null padding is set to false

pdet committed:
Fix for test replacement scan alias

pdet committed:
We can now detect this as a one column string

pdet committed:
Refuse newline detection if does nto fit what was provided by the user

pdet committed:
I don't think the previous test result of this test made sense 5077

pdet committed:
Add unquoted state

pdet committed:
Return -1 on sniffed buffer position since we don't really check the last pos

pdet committed:
This file is so borked anything can be a decent valid answer

pdet committed:
Count how many lines to skip and skip completely emply lines for CSV Dialect Detection

pdet committed:
Fix python build

pdet committed:
Merge remote-tracking branch 'origin/feature' into new_sniffer

pdet committed:
Fix amalgamation

pdet committed:
Include-fest

pdet committed:
Don't write empty quotes and escapes into the csv file

pdet committed:
Fix col number

pdet committed:
Reverting to dialect sniffing first and doing some cleanup

pdet committed:
Adding the CSV Buffer Manager

pdet committed:
Fixing minor issues

pdet committed:
Adding regressions tests for CSV Reader

pdet committed:
Includes for windows'

pdet committed:
Adding new escape quote option

pdet committed:
Malloc smaller buffers stop refinement when having one candidate only

pdet committed:
Fix Issue with finished on compressed files and line size checking

pdet committed:
Merge remote-tracking branch 'origin' into csv_sniffer

pdet committed:
Format

pdet committed:
Decoupling sniffer from Buffered CSV Reader Part I

pdet committed:
More on the new type detection

pdet committed:
Seems like the very basics of type detection are working

pdet committed:
Use options, remove specialized option of csv state

pdet committed:
Break sniffer into more manageable files

pdet committed:
UTF-8 Checking of the CSV strings, separating header for header detection, error checking

pdet committed:
Adding Header Detection, removing Candidates struct, using state machines directly

pdet committed:
Adding type replacement and some cleanup

pdet committed:
Type refinement

pdet committed:
Type refinement

pdet committed:
Out when finishing buffer in type refinement

pdet committed:
Properly parsing quotes values, fix bug with producing types

pdet committed:
Fix for date format detection

pdet committed:
Properly initialize buffered reader from read_csv_auto bind

pdet committed:
Localize dialect detection variables

pdet committed:
Properly set parse chunk size and refine date formats

pdet committed:
Skip lines before starting buffer manager, check for set header

pdet committed:
Use user defines chunk size

pdet committed:
Treat empty value in csv as null

pdet committed:
Null value

pdet committed:
Making sure to run the sniffer for every file on union by name

pdet committed:
Parse values into chunks fix and run sniffer in copy csv

pdet committed:
Skip empty lines

pdet committed:
Properly run the sniffer from copy_csv

pdet committed:
Skip notes on header check for BOM

pdet committed:
Format and error message fix

pdet committed:
Pedantic was not happy with extra ;

pdet committed:
Copy type to avoid OOS

pdet committed:
Make csv_quoted_newline_incorrect.test pass by skipping malformed lines

pdet committed:
Fix benchmark name

pdet committed:
Properly evaluate correct number of values, fix py test

pdet committed:
Includes to fix windows build

pdet committed:
Fix for union

pdet committed:
CSV state machine handling null paddling

pdet committed:
Run sniffer in relation

pdet committed:
Missing include for windows

pdet committed:
Fix serialization issue, error message and read_csv options

pdet committed:
remove accidental debug code

pdet committed:
Pass csv types of sniffer in buffered constructor

pdet committed:
Merge remote-tracking branch 'origin/master' into csv_sniffer

pdet committed:
enum merge

pdet committed:
Check if iterator is valid first

pdet committed:
WIndows empty lines

pdet committed:
Auto generate for enumM

carlopi commented on src/include/duckdb/execution/operator/persistent/csv_scanner/base_csv_reader.hpp:
```diff
@@ -30,19 +30,19 @@ struct StrpTimeFormat;
 class FileSystem;
 
-enum class ParserMode : uint8_t { PARSING = 0, SNIFFING_DIALECT = 1, SNIFFING_DATATYPES = 2, PARSING_HEADER = 3 };
+enum class ParserMode : uint8_t { PARSING = 0, SNIFFING_DATATYPES = 1, PARSING_HEADER = 2 };
```
Small note here, independent but inspired from this PR: it would be useful for stable (de)serialisation across version to have a way to signify: this enum value is to be skipped AND should not be reused any more. Otherwise removing / renumbering enums might results in incompatibilities between versions.

pdet committed:
Ghetto fixing windows empty lines

pdet committed:
Fix error message

pdet committed:
Maximum line size fix for buffered reader

pdet committed:
Fixing issue with calculating line size used in buffer test

pdet committed:
Force disable csv buffer in this PR

pdet committed:
Adding tests for #8320, move max line size test to slow, increase minimum buffer size allocation

pdet committed:
Merge remote-tracking branch 'origin/master' into csv_sniffer

pdet committed:
Serializer updates

pdet committed:
Json deserializer is unhappy:
|

taniabogatsch commented:
Oh wow! I (GitHub stale bot) found this issue and thought it might be related: https://github.com/duckdb/duckdb/issues/5250.

pdet committed:
Merge remote-tracking branch 'origin/master' into csv_sniffer

pdet committed:
Test for issue #5250

pdet committed:
Have to write chars as well

pdet committed:
Initial wrapper of the FileBuffers and pin unpin of csv files

pdet committed:
More on pin unpin buffers

pdet committed:
Pin Unpin seems solid, but something is wrong with the CSV Buffer

pdet committed:
Pin/Unpin final

pdet committed:
Template and wrap process functions over buffer iterator

pdet committed:
More on dialect new loop sniffing operator

pdet committed:
type detection on the templated loop

pdet committed:
Parsing Detection on the new loop

pdet committed:
Caching for CSV State Machines

pdet committed:
Avoid the mumbo-jumbo of csv copying

pdet committed:
More on optimization for small csv files

pdet committed:
More on small file optimization

pdet committed:
construct on first use csv state machine cache

pdet committed:
Missing include

pdet committed:
Small fixes

pdet committed:
pedantic

pdet committed:
One more optimization for dialect detection

pdet committed:
Headers to make windows happy and small fix

pdet committed:
Update parallel csv reader to work with new buffer system

pdet committed:
remove static variable, move cache to client data and run the format serializer thing

pdet committed:
Fix serialization nodes

pdet committed:
Merge remote-tracking branch 'origin/master' into csv_sniffer

taniabogatsch commented on src/include/duckdb/execution/operator/persistent/csv_scanner/csv_state_machine_cache.hpp:
```diff
@@ -0,0 +1,30 @@
+#include "duckdb/execution/operator/persistent/csv_scanner/csv_buffer_manager.hpp"
+
+namespace duckdb {
+typedef uint8_t state_machine_t[7][256];
```
It's a bit confusing to have the type definition of the state machine outside of the `csv_state_machine.hpp`. And I think we want to make `7` (and potentially `256`) `static constexpr`, so we can change them in a single place (probably `CSVStateMachine`) and don't have 'magic' numbers around.

taniabogatsch commented on src/include/duckdb/execution/operator/persistent/csv_scanner/csv_state_machine.hpp:
```diff
@@ -0,0 +1,78 @@
+	string value;
+	idx_t rows_read;
+
+	//! Options resulting from sniffing
+	char quote;
+	char escape;
+	char delim;
+	NewLineIdentifier new_line = NewLineIdentifier::NOT_SET;
+	idx_t num_cols = 0;
+	bool header = false;
+	std::map<LogicalTypeId, bool> has_format;
+	std::map<LogicalTypeId, StrpTimeFormat> date_format;
+	idx_t skip_rows = 0;
```
What makes these different from the `CSVReaderOptions`? Can we not reuse them? Or even a subset of them?
```cpp
const CSVReaderOptions &options;
CSVReaderOptions sniffing_options;
```

taniabogatsch commented on src/include/duckdb/execution/operator/persistent/csv_scanner/csv_state_machine_cache.hpp:
```diff
@@ -0,0 +1,30 @@
+private:
+	void Insert(char delimiter, char quote, char escape);
+	//! Cache on delimiter|quote|escape
+	unordered_map<char, unordered_map<char, unordered_map<char, state_machine_t>>> state_machine_cache;
```
I am curious about this. It is a bit hard to read, is there a significant performance reason behind not using a `struct`? If so, then we should add a comment. 

Otherwise, can we not have a struct and add a custom equal operator? Something like this:
```cpp
struct CSVStateMachineConfig {
char delimiter;
char quote;
char escape;
};
```

pdet replied:
With a struct we have O(N) access no? Not that I expect the cache to get ginormous. I like the O(1) and I don't think is that bad to read

taniabogatsch replied:
`unordered_map<my_struct, state_machine_t>` with a custom comparison function and hash function on the struct should still be `O(1)`, no? I was thinking of something like this, but I could be mistaken. :) It is more extendable (if we get more state machine options, not sure about that). But I can also see that it adds more code—just my thoughts. 😄 

```cpp
struct CSVStateMachineConfig {
	char delimiter;
	char quote;
	char escape;

	bool operator==(const CSVStateMachineConfig &other) const {
	    return delimiter == other.delimiter && quote == other.quote && escape == other.escape;
	}
};

// possibly want to use our hash function (duckdb/common/types/hash.hpp)
struct HashCSVStateMachineConfig {
	std::size_t operator()(CSVStateMachineConfig const& config) const noexcept {
		auto h_delimiter = std::hash<char>{}(config.delimiter);
		auto h_quote = std::hash<char>{}(config.quote);
		auto h_escape = std::hash<char>{}(config.escape);
		return CombineHash(h_delimiter, CombineHash(h_quote, h_escape));
	}
};

unordered_map<CSVStateMachineConfig, state_machine_t> configurations;
```

pdet replied:
fair enough :-)

taniabogatsch reviewed (changes requested):
Hey @pdet! Really impressive work! 😲 

As this is a huge PR, I will split my review into sub-reviews. In my first review, I mostly went through all the different files to get a better overview of your changes in general and their impact. I also checked off minor changes to test files, serialization, etc.

Then I started looking into a few header files, such as `csv_state_machine.hpp`, and left some comments. I tried to evaluate how 'clear' the header files are without necessarily looking at the function implementations.

Next, I'll have a go at an in-depth look at the `cpp` files. Cheers! 😄

taniabogatsch commented on src/include/duckdb/execution/operator/persistent/csv_scanner/csv_buffer.hpp:
```diff
@@ -0,0 +1,102 @@
+#include "duckdb/storage/buffer/block_handle.hpp"
+
+namespace duckdb {
+static constexpr idx_t CSV_BUFFER_SIZE = 32000000; // 32MB
+//! In case the file has a size < 32MB, we will use this size instead
+//! This is to avoid mallocing a lot of memory for a small file
+//! And if it's a compressed file we can't use the actual size of the file
+static constexpr idx_t CSV_MINIMUM_BUFFER_SIZE = 10000000; // 10MB
```
Do they need to be global? Or can we move them into the `CSVBuffer` class? You only use it in `csv_buffer_manager.cpp` and `csv_reader_options.hpp`, and I think we usually keep our `constexpr` inside the context of structs or classes.

taniabogatsch commented on src/include/duckdb/execution/operator/persistent/csv_scanner/csv_buffer_manager.hpp:
```diff
@@ -0,0 +1,105 @@
+namespace duckdb {
+class CSVBuffer;
+class CSVStateMachine;
+//! This class is used to manage the buffers
+//! Buffers are cached when used for auto detection
+//! Otherwise they are not cached and just returned
```
What are these buffers, and what's in them? Parts of a CSV file? Let's extend the class description a bit. :)

taniabogatsch commented on src/include/duckdb/execution/operator/persistent/csv_scanner/csv_buffer_manager.hpp:
```diff
@@ -0,0 +1,105 @@
+	                 bool cache_buffers = true);
+	//! Returns a buffer from a buffer id (starting from 0). If it's in the auto-detection then we cache new buffers
+	//! Otherwise we remove them from the cache if they are already there, or just return them bypassing the cache.
+	unique_ptr<CSVBufferHandle> GetBuffer(idx_t pos, bool auto_detection);
```
Nit: `unique_ptr<CSVBufferHandle> GetBuffer(const idx_t pos, const bool auto_detection = false);` maybe?

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/type_replacement.cpp:
```diff
@@ -0,0 +1,39 @@
+			}
+			if (!best_candidate->options.file_options.union_by_name &&
+			    found < best_candidate->options.sql_types_per_column.size()) {
+				string exception = BufferedCSVReader::ColumnTypesError(options.sql_types_per_column, names);
```
Nit: rename `exception` to `error_msg`?

taniabogatsch commented on src/execution/operator/csv_scanner/parallel_csv_reader.cpp:
```diff
@@ -592,9 +585,13 @@ final_state : {
+					VerifyLineLength(position_buffer - line_start, buffer->batch_index);
 					line_start = position_buffer;
 					AddRow(insert_chunk, column, error_message, buffer->local_batch_index);
+					if (position_buffer - verification_positions.end_of_last_line > options.buffer_size) {
```
Same as before:
> Is this passed to the user? If so, maybe it should contain a bit more of an explanation. From my understanding, we plan to move to use the parallel CSV reader by default (or we are already doing that). Then, if users may do FROM 'my_file.csv, receiving this error message would be pretty confusing, no?

taniabogatsch commented on src/execution/operator/csv_scanner/base_csv_reader.cpp:
From my understanding of what's happening here: We throw out a lot of code related to our previous sniffing implementation. Is that correct?

pdet replied:
Yeah, basically the sniffer used to be quite ingrained to the buffered csv reader. For clarification, a lot of the old sniffer was not completely removed but recycled :-)

taniabogatsch replied:
Nice! :) It's also neat that this PR removes much more code than it adds, and it's not only multi-character stuff that's being removed. 🎉

taniabogatsch reviewed (changes requested):
I continued with the review 😅 and started tackling the sniffer function implementations. To be continued tomorrow!

pdet commented on .github/workflows/Regression.yml:
```diff
@@ -119,6 +119,12 @@ jobs:
+      if: always()
+      shell: bash
+      run: |
+        python scripts/regression_test_runner.py --old=duckdb/build/release/benchmark/benchmark_runner --new=build/release/benchmark/benchmark_runner --benchmarks=.github/regression/csv.csv --verbose --threads=2
```
These regression tests are made to be executed with the CI, which only has [2 cores](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners)

taniabogatsch replied:
Ah, thanks for the reference!

pdet committed:
Don't really need the cranlogs.benchmark

pdet committed:
Add proper path to small csv test

pdet commented on src/include/duckdb/execution/operator/persistent/csv_scanner/csv_state_machine.hpp:
```diff
@@ -0,0 +1,78 @@
+
+//! All States of CSV Parsing
+enum class CSVState : uint8_t {
+	STANDARD = 0,         //! Regular unquoted field state
+	FIELD_SEPARATOR = 1,  //! State after encountering a field separator (e.g., ;)
```
Your suggestion then is to replace `STANDARD` with `FIELD` and `FIELD_SEPARATOR` with `DELIMITER`?

taniabogatsch replied:
`DELIMITER` looks good. Let's keep `STANDARD`.

pdet committed:
Unify delimiter name

pdet commented on src/include/duckdb/main/client_data.hpp:
```diff
@@ -62,6 +63,8 @@ struct ClientData {
 	//! The file search path
 	string file_search_path;
 
+	CSVStateMachineCache state_machine_cache;
```
I think these are different. The `debug_max_line_length` is a flag only really used for the small buffer size tests from `test/parallel_csv/test_parallel_csv.cpp` so I do think is a bit dirty and should be removed.

However the `CSVStateMachineCache`is basically a cache of the state machines that are frequently used in that session, so I think it makes sense to have in the `ClientData`, but I'm ofc open for suggestions.

taniabogatsch replied:
My concern about this is that, from my understanding, we keep the state machine cache around for whatever we do in a client session. I.e., we might use it for a query reading from CSV files, but it is not required for any other queries.

I assume we've one `ClientData` struct for a connection (`client_context.hpp`).
```cpp
	//! The set of client-specific data
	unique_ptr<ClientData> client_data;
```	

And yes, the `FIXME` should go. 😅

taniabogatsch replied:
But I get your point of multiple (repeated) CSV reads in one session without creating a table from the file(s). If we are going for this, I'd make it more transparent, possibly with a descriptive comment on why we keep these state machines around.

pdet committed:
Add descriptive comment of what a CSVStateMachine is and fix minor comment errors about the new line states

pdet committed:
Remove magic numbers

pdet committed:
Creating struct for dialect detection options

pdet committed:
Finalize struct

pdet committed:
More of Tania's comments

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/header_detection.cpp:
```diff
@@ -0,0 +1,165 @@
+		}
+	}
+
+	for (idx_t i = 0; i < MinValue<idx_t>(names.size(), best_candidate->options.name_list.size()); i++) {
+		names[i] = best_candidate->options.name_list[i];
+	}
```
What is happening here?

pdet replied:
It replaces names discovered from the sniffer with names defined by the user. I'm adding a comment.

taniabogatsch reviewed (changes requested):
I just went over it a bit more. But I'll wait to look at some files until you've extended the state machine to support new line detection. :) So let me know how that goes!

pdet committed:
format

pdet committed:
Use our own HF in the CSV State Machine cache

pdet committed:
Lot of small PR requests

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/dialect_detection.cpp:
```diff
@@ -0,0 +1,311 @@
+
+		machine.state = static_cast<CSVState>(
+		    machine.transition_array[static_cast<uint8_t>(machine.state)][static_cast<uint8_t>(current_char)]);
+		bool empty_line =
+		    (machine.state == CSVState::CARRIAGE_RETURN && machine.previous_state == CSVState::CARRIAGE_RETURN) ||
+		    (machine.state == CSVState::RECORD_SEPARATOR && machine.previous_state == CSVState::RECORD_SEPARATOR) ||
+		    (machine.state == CSVState::CARRIAGE_RETURN && machine.previous_state == CSVState::RECORD_SEPARATOR) ||
+		    (machine.pre_previous_state == CSVState::RECORD_SEPARATOR &&
+		     machine.previous_state == CSVState::CARRIAGE_RETURN);
```
If within reasonable effort, we should have this as another state in the state machine. We can then skip `// It means our carriage return is a record separator` and `// Identify what is our line separator`, as that should happen when we try to find the best state machine (I think 😄). `Finalize` should also become simpler.

pdet committed:
More small PR requests

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/dialect_detection.cpp:
```diff
@@ -0,0 +1,311 @@
+		}
+	}
+
+	if (num_cols < prev_column_count) {
+		// Early return if we have less columns than the previous chunk run
+		return;
+	}
```
Is this for simplicity? If possible, does null-padding to the maximum column count add unnecessary complexity?

pdet replied:
In case padding is not accepted?

taniabogatsch replied:
Oh yea, true, I missed that.

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/dialect_detection.cpp:
```diff
@@ -0,0 +1,311 @@
+		return;
+	}
+
+	// some logic
```
😅

pdet committed:
Merge remote-tracking branch 'origin/main' into csv_sniffer

pdet committed:
Finishing the merge of the nodes

pdet committed:
New Empty Line state

pdet commented:
Hi @taniabogatsch, thanks again for the thorough review! I've gone ahead and addressed all the comments I marked as resolved. I've left some unresolved, in case you still want to discuss my answers. I've also added a new state for empty lines.

If you're up for it, another pass would be great. Otherwise, I'll be wrestling with CI for the next few days.

taniabogatsch commented on src/execution/operator/csv_scanner/parallel_csv_reader.cpp:
```diff
@@ -452,6 +440,10 @@ add_row : {
 		}
 		SkipEmptyLines();
+		if (position_buffer - verification_positions.end_of_last_line > options.buffer_size) {
+			error_message = "Line does not fit in one buffer. Increase the buffer size.";
```
Okay, that makes sense. At that point, we assume that the user knows that they have massive lines in their CSV file :) Is the 32MB threshold documented transparently with the option to increase the buffer size?

taniabogatsch replied:
A quick Google search did not lead to anything for me, except this: https://github.com/duckdb/duckdb/discussions/8102

pdet replied:
@szarnyasg This is something we should indeed consider documenting, possibly after merging this PR. The current rule (in this PR) is that by default, buffers will have 32Mb, unless the file is smaller than 32Mb (In that case it will have the same size as the file size). The `buffer_size` option then allows users who know what they are doing to override this.

taniabogatsch commented:
Haha, sorry for sneaking in these few more comments. 😄  I don't think that they'll introduce large changes, so I'll do another pass now while you wrestle with the CI 💪

pdet committed:
Fix for #8649

pdet committed:
Move state machine cache code to a proper cpp file

pdet committed:
Simplify state machine initalization

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/dialect_detection.cpp:
```diff
@@ -0,0 +1,307 @@
+
+	if (!requested_types.empty() && requested_types.size() != num_cols && !invalid_padding) {
+		return;
+	} else if (rows_consistent &&
```
Nit: remove `else if` here and in the following case, and replace with `if` that returns for readability.

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/dialect_detection.cpp:
```diff
@@ -0,0 +1,307 @@
+	bool start_good = !candidates.empty() && (start_row <= candidates.front()->start_row);
+	bool invalid_padding = !allow_padding && padding_count > 0;
+
+	if (!requested_types.empty() && requested_types.size() != num_cols && !invalid_padding) {
```
The three cases of this if-statement have quite long conditions. Can we have comments to describe the grouping/reasoning behind these three cases?

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/dialect_detection.cpp:
```diff
@@ -0,0 +1,307 @@
+			bool finished_file = cur_candidate->csv_buffer_iterator.Finished();
+			if (finished_file || i == options.sample_chunks) {
+				// we finished the file or our chunk sample successfully: stop
+				auto neo = std::move(cur_candidate);
```
Nice reference. 😄 But can we rename it to `successful_candidate`?

pdet committed:
Move quote_rules to its hpp, comments on dialect detection, assertions

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/type_detection.cpp:
```diff
@@ -0,0 +1,407 @@
+#include "duckdb/common/operator/decimal_cast_operators.hpp"
+#include "duckdb/execution/operator/persistent/csv_scanner/csv_sniffer.hpp"
+#include <algorithm>
+#include <string>
```
Nit: `duckdb/common/algorithm.hpp` and `duckdb/common/string`

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/type_detection.cpp:
```diff
@@ -0,0 +1,407 @@
+		for (idx_t i = 0; i < candidate->dialect_options.num_cols; i++) {
+			info_sql_types_candidates[i] = candidate->options.auto_type_candidates;
+		}
+		std::map<LogicalTypeId, bool> has_format_candidates;
+		std::map<LogicalTypeId, vector<string>> format_candidates;
```
Nit: `duckdb/common/map.hpp`

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/type_detection.cpp:
```diff
@@ -0,0 +1,407 @@
+									if (format_string.find("%Y-%m-%d") == string::npos) {
+										type_format_candidates.emplace_back(format_string);
+									}
+								}
+							}
+							//	initialise the first candidate
+							candidate->dialect_options.has_format[sql_type.id()] = true;
+							//	all formats are constructed to be valid
+							SetDateFormat(*candidate, type_format_candidates.back(), sql_type.id());
+						}
+						// check all formats and keep the first one that works
+						StrpTimeFormat::ParseResult result;
+						auto save_format_candidates = type_format_candidates;
+						while (!type_format_candidates.empty()) {
+							//	avoid using exceptions for flow control...
+							auto &current_format = candidate->dialect_options.date_format[sql_type.id()];
+							if (current_format.Parse(StringValue::Get(dummy_val), result)) {
+								break;
+							}
+							//	doesn't work - move to the next one
+							type_format_candidates.pop_back();
+							candidate->dialect_options.has_format[sql_type.id()] = (!type_format_candidates.empty());
+							if (!type_format_candidates.empty()) {
+								SetDateFormat(*candidate, type_format_candidates.back(), sql_type.id());
+							}
+						}
+						//	if none match, then this is not a value of type sql_type,
+						if (type_format_candidates.empty()) {
+							//	so restore the candidates that did work.
+							//	or throw them out if they were generated by this value.
+							if (had_format_candidates) {
+								type_format_candidates.swap(save_format_candidates);
+								if (!type_format_candidates.empty()) {
+									SetDateFormat(*candidate, type_format_candidates.back(), sql_type.id());
+								}
+							} else {
+								has_format_candidates[sql_type.id()] = false;
+							}
+						}
+					}
```
Can we move this into a separate function? I did not try to follow this logic; I shy away from anything date/time-related. 😅

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/type_refinement.cpp:
```diff
@@ -0,0 +1,187 @@
+	}
+};
+
+bool CSVSniffer::TryCastVector(Vector &parse_chunk_col, idx_t size, const LogicalType &sql_type) {
```
We should use early-outs here for readability.

taniabogatsch commented on src/execution/operator/csv_scanner/sniffer/type_refinement.cpp:
```diff
@@ -0,0 +1,187 @@
+		} else {
+			detected_types = requested_types;
+		}
+	} else if (best_candidate->options.all_varchar) {
+		// return all types varchar
+		return;
```
This should be an early-out.

taniabogatsch reviewed (changes requested):
Hey @pdet! I just finished my second pass over your PR, and it is nice to see all the changes in place! I added more comments, but they're primarily minor remarks on readability.

Also, while reviewing your changes, I wondered about the following points.
- I hope that I found all `std::` occurrences for which we have our own headers
- If you have the time, then it might be worth going through the header files and adding `const` to eligible function parameters and function definitions. I found this especially useful for the `CSVReaderOptions` so that I know in advance if I can expect modifications on these from a function.

As for benchmarks/regressions, I think that we should have the following:
- many small CSV files with [different|same] dialect
- a few huge CSV files with [different|same] dialect

In the PR I only see a benchmark for `lineitem` and one for a singular small CSV file.

Cheers 🙌

pdet commented on benchmark/micro/csv/sniffer.benchmark:
Well, the difference is that describe only runs the sniffer, but it doesn't actually parse the CSV file.
While a read indeed also runs the sniffer, but has the added cost of parsing, constructing duckdb chunks, projection yadda yadda yadda

taniabogatsch replied:
Ah, that makes sense! Thanks.

pdet commented on src/execution/operator/csv_scanner/buffered_csv_reader.cpp:
```diff
@@ -0,0 +1,530 @@
+}
+
+void BufferedCSVReader::SkipRowsAndReadHeader(idx_t skip_rows, bool skip_header) {
+	for (idx_t i = 0; i < skip_rows; i++) {
+		// ignore skip rows
+		string read_line = file_handle->ReadLine();
+		linenr++;
+	}
+
+	if (skip_header) {
+		// ignore the first line as a header line
+		InitParseChunk(return_types.size());
+		ParseCSV(ParserMode::PARSING_HEADER);
+	}
```
Ha, some csv magic for you, we need to be able to read this kind of stuff:
```
some notes...
more notes,...
number,text,date
345.0,TEST4,2000-12-13
```

I think they are actually frequently connected., and splitting them would potentially add more confusion

taniabogatsch replied:
Uh oh. I did not think of that/know about it. Makes total sense now.

pdet committed:
Favor map instead of std::map, remove unused functions, refactor variable names

pdet commented on src/execution/operator/csv_scanner/csv_buffer.cpp:
```diff
@@ -0,0 +1,90 @@
+	auto buffer = Ptr();
+	actual_size = file_handle.Read(buffer, buffer_size_p);
+	global_csv_start = global_csv_current_position;
+	if (actual_size >= 3 && buffer[0] == '\xEF' && buffer[1] == '\xBB' && buffer[2] == '\xBF') {
+		start_position += 3;
+	}
```
jup, adding comment

pdet committed:
Adding more comments and early out instead of nested ifs

pdet committed:
Create state_machine_options to hold delimiter, quote and escape chars

pdet committed:
test nits

pdet committed:
Nits and comments

pdet committed:
nips and format

pdet committed:
More early outs and modularization of date timestamp format detection

pdet committed:
Python Fix

pdet commented on src/function/table/read_csv.cpp:
```diff
@@ -1090,35 +1088,74 @@ void CSVComplexFilterPushdown(ClientContext &context, LogicalGet &get, FunctionD
+	writer.WriteList<string>(csv_formats);
+}
+
+void DialectOptions::Deserialize(FieldReader &reader) {
```
If that's the case, I think it is way beyond the scope of this PR.
These are basically compatible with the LogicalOperator Serialize and Deserialize.

pdet committed:
Add benchmarks

pdet committed:
Small bug related to \n\r not going to an empty line

pdet committed:
Fixing bug related to not finding dialect candidates

pdet committed:
Integrating Buffer manager with buffered csv reader

pdet committed:
random step in the right direction

carlopi commented on src/include/duckdb/storage/serialization/nodes.json:
```diff
@@ -457,190 +457,155 @@
+      },
+      {"id": 135,
+        "name": "dialect_options.has_format",
+        "type": "unordered_map<bool, StrpTimeFormat> date_format"
```
Same, remove date_format since only type is needed

carlopi reviewed (commented):
Hi! Went through the code, nice to read. Left punctual comments in a couple of places.

pdet committed:
More on parallel csv reader using buffer manager

pdet committed:
More of buffer manager

pdet committed:
Parallel csv reader can take empty options

pdet committed:
More on the buffer manager

pdet committed:
small adjustments

pdet committed:
Small adjustments

pdet commented on src/include/duckdb/execution/operator/persistent/csv_scanner/csv_state_machine_cache.hpp:
```diff
@@ -0,0 +1,51 @@
+namespace duckdb {
+static constexpr uint32_t NUM_STATES = 8;
+static constexpr uint32_t NUM_TRANSITIONS = 256;
+typedef uint8_t state_machine_t[NUM_STATES][NUM_TRANSITIONS];
```
A state machine should be around 2048 bytes (unless states get added and such). I think for most cases the cache will basically contain ~ 18 options (our default cases), so I guess is not worth the added complexity to compress this yet.

pdet committed:
Carlo's comments

pdet commented on src/execution/operator/csv_scanner/csv_reader_options.cpp:
```diff
@@ -60,60 +60,69 @@ static int64_t ParseInteger(const Value &value, const string &loption) {
-		new_line = NewLineIdentifier::CARRY_ON;
+		dialect_options.new_line = NewLineIdentifier::CARRY_ON;
 	} else {
 		throw InvalidInputException("This is not accepted as a newline: " + input);
```
Agree, I will keep that in mind when refactoring the Parallel CSV Reader

pdet commented on src/include/duckdb/execution/operator/persistent/csv_scanner/csv_state_machine.hpp:
```diff
@@ -0,0 +1,77 @@
+	idx_t start_row = 0;
+	//! The Transition Array is a Finite State Machine
+	//! It holds the transitions of all states, on all 256 possible different characters
+	state_machine_t &transition_array;
```
The cache doesn't remove anything, and it only gets borked at the end of the session, so that should be fine I think.

pdet committed:
Format and input sanitizing

pdet committed:
Merge remote-tracking branch 'origin/main' into csv_sniffer

pdet committed:
Pedro Versus the CI Machines

pdet committed:
Pedro Versus the CI Machines

pdet committed:
Pedro Versus the CI Machines: R can't handle over 100 byte includes

pdet committed:
Run Parallel on single file

pdet committed:
make formatter happy

pdet committed:
CodeCov: Buffer Manager Finalize Test

pdet committed:
Buffer can't be nullptr in CSVBufferRead

pdet committed:
Force test on reload

pdet committed:
Invalid UTF test

pdet committed:
Remove unused code

pdet committed:
Update uncovered lines with new file paths

pdet committed:
Fixing code and tests for blob checking

pdet committed:
100% codecov in dialect detection

pdet committed:
100% codecov in header detection

pdet committed:
Removing sniffing code for decimal and float detection, 100% cov in sniffing types

pdet committed:
Type Refinement 100% codecov and Format

pdet committed:
Merge remote-tracking branch 'origin/main' into csv_sniffer

pdet committed:
Finish the merge with main, was more complex than I thought

Mytherin commented on test/sql/copy/csv/auto/test_auto_cranlogs.test:
```diff
@@ -1,34 +0,0 @@
-# group: [auto]
-
-statement ok
-CREATE TABLE cranlogs AS SELECT * FROM read_csv_auto ('test/sql/copy/csv/data/real/tmp2013-06-15.csv.gz');
```
Can you add this test back in?

Mytherin reviewed (commented):
Thanks for the massive PR! Looks great - some comments below:

pdet committed:
Re-adding cran test and removing unnecessary pending query test

pdet committed:
Moving state machine cache

pdet committed:
more on codecov

pdet committed:
more on codecov

pdet committed:
Can safely remove this, it's now handled by the buffers with destroy settings

pdet committed:
Updating path

pdet committed:
Reduce SF of micro benchmark so CI can reliably finish it

pdet committed:
wrong path

pdet commented:
@Mytherin , is this good to go?
